src/crypto.rs

```diff
@@ -1,158 +1,240 @@
 use std::convert::TryFrom;
 
-use aes::{Aes128, Aes256};
-use block_modes::block_padding::Pkcs7;
-use block_modes::BlockMode;
-use cipher::generic_array::typenum::{U16, U32};
-use cipher::generic_array::{ArrayLength, GenericArray};
-use cipher::NewBlockCipher;
+use aes::cipher::block_padding::Pkcs7;
+use aes::cipher::{BlockDecryptMut, BlockEncryptMut, KeyIvInit};
+use cipher::generic_array::GenericArray;
 use rand::{thread_rng, RngCore};
 
 use crate::{ErrorInfo, Result};
 
-/// A 128 bit AES key.
-#[derive(Clone)]
-pub struct Key128(GenericArray<u8, U16>);
+type Aes128CbcEnc = cbc::Encryptor<aes::Aes128>;
+type Aes256CbcEnc = cbc::Encryptor<aes::Aes256>;
+type Aes128CbcDec = cbc::Decryptor<aes::Aes128>;
+type Aes256CbcDec = cbc::Decryptor<aes::Aes256>;
 
-/// A 256 bit AES key.
-#[derive(Clone)]
-pub struct Key256(GenericArray<u8, U32>);
+pub(crate) type IV = [u8; 16];
 
-// Type alias for CBC mode with Pkcs7 padding.
-type Cbc<T> = block_modes::Cbc<T, Pkcs7>;
-
-// An internal type for providing a deterministic IV in tests.
-pub(crate) type IV = [u8; aes::BLOCK_SIZE];
-
-#[derive(Clone)]
-pub enum Key {
+#[derive(Clone, Debug)]
+pub enum Cipher {
     /// A 128 bit AES key.
-    Key128(Key128),
+    Aes128Cbc([u8; 16]),
     /// A 256 bit AES key.
-    Key256(Key256),
+    Aes256Cbc([u8; 32]),
+}
+
+impl Default for Cipher {
+    fn default() -> Self {
+        Self::builder().build().unwrap()
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
+pub enum CipherKind {
+    AesCbc,
+}
+
+impl Default for CipherKind {
+    fn default() -> Self {
+        CipherKind::AesCbc
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
+pub enum KeyLen {
+    Bits128,
+    Bits256,
+}
+
+#[derive(Clone, Debug, Default)]
+pub struct CipherBuilder {
+    kind: CipherKind,
+    len: Option<KeyLen>,
+    key: Option<Vec<u8>>,
+}
+
+impl CipherBuilder {
+    pub fn kind(mut self, kind: CipherKind) -> Self {
+        self.kind = kind;
+        self
+    }
+
+    pub fn key(mut self, key: Vec<u8>) -> Self {
+        self.key = Some(key);
+        self
+    }
+
+    pub fn string(mut self, key: &str) -> Result<Self> {
+        let key = base64::decode(key)?;
+        self.key = Some(key);
+        Ok(self)
+    }
+
+    pub fn len(mut self, len: KeyLen) -> Self {
+        self.len = Some(len);
```

Comment:
A cipher always needs a key, so can the length be determined from whatever key is passed in? And in fact, is `KeyLen` not required because we could just set `self.cipher = Aes128Cbc | Aes256Cbc` when the user passes the key?

Reply:
The spec had some other function to generate a random key where you specified the size. And then you had to put a key into the function to get cipher params. 

I saw no reason for two functions to merged them into one. If you don't specify a key then one will be generated.

Reply:
> If you don't specify a key then one will be generated.

Ok but if one gets generated, the user needs to know what it is so they can decrypt the message later, so the separate functions make that a bit more explicit (i.e. first you generate a key, and perhaps store it in a database, then you use that key each time you want to encrypt/decrypt data on a channel).

Reply:
Like I said below there could be a convenience function to do it easier but the key data is not private.